src/rust/engine/src/downloads.rs

```diff
@@ -109,25 +124,72 @@ impl StreamingDownload for FileDownload {
   }
 }
 
-async fn start_download(
+async fn attempt_download(
   core: &Arc<Core>,
-  url: Url,
+  url: &Url,
   file_name: String,
-) -> Result<Box<dyn StreamingDownload>, String> {
-  if url.scheme() == "file" {
-    if let Some(host) = url.host_str() {
-      return Err(format!(
-        "The file Url `{}` has a host component. Instead, use `file:$path`, \
-        which in this case might be either `file:{}{}` or `file:{}`.",
-        url,
-        host,
-        url.path(),
-        url.path(),
-      ));
+  expected_digest: Digest,
+) -> Result<(Digest, Bytes), StreamingError> {
+  let mut response_stream: Box<dyn StreamingDownload> = {
+    if url.scheme() == "file" {
+      if let Some(host) = url.host_str() {
+        return Err(StreamingError::Permanent(format!(
+          "The file Url `{}` has a host component. Instead, use `file:$path`, \
+          which in this case might be either `file:{}{}` or `file:{}`.",
+          url,
+          host,
+          url.path(),
+          url.path(),
+        )));
+      }
+      Box::new(FileDownload::start(url.path(), file_name).await?)
+    } else {
+      Box::new(NetDownload::start(core, url.clone(), file_name).await?)
     }
-    return Ok(Box::new(FileDownload::start(url.path(), file_name).await?));
+  };
+
+  struct SizeLimiter<W: std::io::Write> {
+    writer: W,
+    written: usize,
+    size_limit: usize,
   }
```

Comment:
Would it make sense to define this outside of the function? This is a big function atm. Ditto the trait.

Reply:
Agreed that it is large. I _think_ that I prefer inlining these single use traits like this (because it makes it clear that no consideration has been made for it being reusable), but I probably would have written it as reusable if I had written it in the first place.